spec/p2p/v0.34/pex-protocol.md

```diff
@@ -0,0 +1,240 @@
+received from it.
+
+Sending a PEX response to a peer that has not requested peer addresses
+is also considered a misbehavior.
+So, if a PEX response is received from a peer that is not registered in
+the `requestsSent` set, a `ErrUnsolicitedList` error is produced.
+This leads the peer to be disconnected and marked as a bad peer.
+
+## Providing Addresses
+
+When a node receives a `PexRequest` message from a peer,
+it replies to peer with a `PexAddrs` message.
+
+This message encodes a [random selection of peer addresses](./addressbook.md#random-selection) 
+retrieved from the address book.
+
+Sending a PEX response to a peer is implemented by the `SendAddrs` method of
+the PEX reactor.
+
+### Misbehavior
+
+Requesting peer addresses to often is considered a misbehavior.
+Since node are expected to send PEX requests every `ensurePeersPeriod`,
+the minimum accepted interval between requests from the same peer is set
+to `ensurePeersPeriod / 3`, 10 seconds by default.
+
+The `receiveRequest` method is responsible for verifying this condition.
+The node keeps a `lastReceivedRequests` map with the time of the last PEX
+request received from every peer.
+If the interval between successive requests is less than the minimum accepted
+one, the peer is disconnected and marked as a bad peer.
+An exception is made for the first two PEX requests received from a peer.
+
+> The probably reason is that, when a new peer is added, the two conditions for
+> a node to request peer addresses can be triggered with an interval lower than
+> the minimum accepted interval.
+> Since this is a legit behavior, it should not be punished.
+
+## Seed nodes
+
+A seed node is a node configured to operate in `SeedMode`.
+
+### Crawling peers
+
+Seed nodes crawl the network, connecting to random peers and sending PEX
+requests to them, in order to learn as many peer addresses as possible.
+More specifically, a node operating in seed mode sends PEX requests in two cases:
+
+1. When a outbound peer is added, and the seed node needs more peer addresses,
+   it requests peer addresses to the new peer
+1. Periodically, the `crawlPeersRoutine` sends PEX requests to a random set of
+   peers, whose addresses are registered in the Address Book
+
+The first case also applies for nodes not operating in seed mode.
+The second case replaces the second for regular nodes, as seed nodes do not
+run the `ensurePeersRoutine`, as regular nodes,
+but run the `crawlPeersRoutine`, which is not run by regular nodes.
+
+The `crawlPeersRoutine` periodically, every 30 seconds (`crawlPeerPeriod`),
+starts a new peer discovery round.
+First, the seed node retrieves a random selection of peer addresses from its
+Address Book.
+This selection is produced in the same way as in the random selection of peer
+addresses that are [provided](#providing-addresses) to a requesting peer.
+Are removed from this selection peers that the seed node has crawled recently,
```

Comment:
```suggestion
From this selection peers that the seed node has crawled recently are removed,
```

Reply:
I adopted the suggestion, but with a different format.